File: src/utils/response.js
/**
 * Send error response
 */
function respondWithError(res, code, message) {
  return res.status(code).json({
    error: {
      message: message || 'An error occurred',
      status: code
    }
  });
}

/**
 * Send success response with data
 */
function respondWithJSON(res, code, data) {
  return res.status(code).json({
    data: data,
    success: true,
    status: code
  });
}

/**
 * Send success response with message
 */
function respondWithMessage(res, code, message) {
  return res.status(code).json({
    message: message,
    success: true,
    status: code
  });
}

/**
 * Log database error
 */
function logDatabaseError(operation, error) {
  console.error(`Database ${operation} error:`, {
    message: error.message,
    stack: error.stack,
    timestamp: new Date().toISOString()
  });
}

/**
 * Log API call
 */
function logAPICall(method, path, userId, statusCode) {
  console.log(`[${new Date().toISOString()}] ${method} ${path} - User: ${userId || 'anonymous'} - Status: ${statusCode}`);
}

/**
 * Validate required fields
 */
function validateRequiredFields(data, requiredFields) {
  const missingFields = [];
  
  requiredFields.forEach(field => {
    if (!data[field] || (typeof data[field] === 'string' && data[field].trim() === '')) {
      missingFields.push(field);
    }
  });
  
  if (missingFields.length > 0) {
    throw new Error(`Missing required fields: ${missingFields.join(', ')}`);
  }
}

/**
 * Parse pagination parameters
 */
function parsePagination(query) {
  const page = parseInt(query.page) || 1;
  const limit = parseInt(query.limit) || 10;
  const offset = (page - 1) * limit;
  
  return { page, limit, offset };
}

/**
 * Format currency
 */
function formatCurrency(amount) {
  return new Intl.NumberFormat('th-TH', {
    style: 'currency',
    currency: 'THB'
  }).format(amount);
}

module.exports = {
  respondWithError,
  respondWithJSON,
  respondWithMessage,
  logDatabaseError,
  logAPICall,
  validateRequiredFields,
  parsePagination,
  formatCurrency
};

File: pkg/utils/response.go
package utils

import (
	"encoding/json"
	"log"
	"net/http"

	"loan-money/internal/models"
)

// RespondWithError sends error response and logs it
func RespondWithError(w http.ResponseWriter, code int, message string) {
	log.Printf("API Error: %d - %s", code, message)
	RespondWithJSON(w, code, models.ErrorResponse{Error: message})
}

// RespondWithJSON sends JSON response
func RespondWithJSON(w http.ResponseWriter, code int, payload interface{}) {
	response, err := json.Marshal(payload)
	if err != nil {
		log.Printf("JSON Encoding Error: %v", err)
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte(`{"error":"Failed to encode response"}`))
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	w.Write(response)
}

// LogDatabaseError logs database errors with more context
func LogDatabaseError(operation string, err error) {
	log.Printf("Database Error [%s]: %v", operation, err)
}

// LogAPICall logs API calls with details
func LogAPICall(method, path, userID string, statusCode int) {
	log.Printf("API Call: %s %s - User: %s - Status: %d", method, path, userID, statusCode)
}
